xarray/core/dataarray.py

```diff
@@ -234,30 +271,37 @@ def __init__(self, data, coords=None, dims=None, name=None,
             variable = Variable(dims, data, attrs, encoding, fastpath=True)
 
         # These fully describe a DataArray
-        self._variable = variable
-        self._coords = coords
-        self._name = name
+        self._variable = variable  # type: Variable
+        assert isinstance(coords, OrderedDict)
+        self._coords = coords  # type: OrderedDict[Hashable, Variable]
+        self._name = name  # type: Optional[str]
 
         # TODO(shoyer): document this argument, once it becomes part of the
         # public interface.
         self._indexes = indexes
 
         self._file_obj = None
 
-        self._initialized = True
+        self._initialized = True  # type: bool
 
-    __default = object()
+    __default = utils.ReprObject('__default')
 
-    def _replace(self, variable=None, coords=None, name=__default):
+    def _replace(self, variable: Optional[Variable] = None,
+                 coords=None,
+                 name: Union[str, None, utils.ReprObject] = __default
```

Comment:
Probably best to just stick with `Hashable` for now for `DataArray.name`. I think this falls in the same category as Dataset keys and dimensions (that is, we currently aren't entirely consistent with supporting non-strings but we definitely do, at least in some cases).